src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Call.cs

```diff
@@ -906,6 +907,81 @@ static bool usesReceiver(BoundExpression argument)
             }
         }
 
+        [return: NotNullIfNotNull(nameof(receiver))]
+        BoundExpression? AdjustReceiverForExtensionsIfNeeded(BoundExpression? receiver, Symbol member)
+        {
+            if (receiver is null)
+            {
+                return receiver;
+            }
+
+            Debug.Assert(member is not null);
+            if (receiver is BoundTypeExpression)
+            {
+                return receiver;
+            }
+
+            if (receiver is BoundThisReference)
+            {
+                // PROTOTYPE this was only minimally tested to observe side-effects but needs more work (in CheckValue for instance)
+                // In an extension type, `this.UnderlyingTypeMember` is replaced with `this.<UnderlyingInstanceField>$.UnderlyingTypeMember`
+                Debug.Assert(receiver.Type is not null);
+                if (receiver.Type.IsExtension && !member.ContainingType.IsExtension)
+                {
+                    var thisExtensionType = (SourceExtensionTypeSymbol)receiver.Type;
+                    Debug.Assert(thisExtensionType.UnderlyingInstanceField is not null);
+                    Debug.Assert(thisExtensionType.UnderlyingInstanceField.Type.Equals(member.ContainingType, TypeCompareKind.AllIgnoreOptions));
+
+                    return _factory.Field(receiver, thisExtensionType.UnderlyingInstanceField);
+                }
+            }
+
+            if (member.ContainingType is not { IsExtension: true } memberExtensionType)
+            {
+                return receiver;
+            }
+
+            Debug.Assert(receiver.Type is not null);
+            if (receiver.Type.IsExtension)
+            {
+                return receiver;
+            }
+
+            ArrayBuilder<LocalSymbol>? temps = ArrayBuilder<LocalSymbol>.GetInstance();
+            ArrayBuilder<BoundExpression>? effects = ArrayBuilder<BoundExpression>.GetInstance();
+            if (receiver.Type.IsReferenceType)
+            {
+                receiver = MakeTemp(receiver, temps, effects);
+            }
+            else if (receiver.Type.IsValueType)
+            {
+                receiver = MakeTemp(receiver, temps, effects, RefKind.Ref);
```

Comment:
>RefKind.Ref [](http://example.com/codeflow?start=62&length=11)

We probably should extract ref kind from `receiver`, see how ```VisitArgumentsAndCaptureReceiverIfNeeded``` figures out what ref kind to use. I think we should share the logic.

Reply:
I didn't follow this. `VisitArgumentsAndCaptureReceiverIfNeeded` saves the receiver into a temp when we need multiple re-uses of the receiver (either because we're doing some compound/complex operation or we'll need the receiver a second time as part of an interpolation handler).
But the purpose of the temp here is not multiple re-uses.

Reply:
> I didn't follow this.

Let's sync up offline